src/zjs_aio_k64f.c

```diff
@@ -0,0 +1,331 @@
+                        false);
+}
+
+// Asynchronous Operations
+static ZJS_DECL_FUNC(zjs_aio_pin_read_async)
+{
+    // args: callback
+    ZJS_VALIDATE_ARGS(Z_FUNCTION);
+
+    return aio_pin_read(function_obj,
+                        this,
+                        argv,
+                        argc,
+                        true);
+}
+
+static ZJS_DECL_FUNC(zjs_aio_pin_close)
+{
+    u32_t pin;
+    zjs_obj_get_uint32(this, "pin", &pin);
+    GET_AIO_HANDLE(this, handle);
+
+    aio_free_cb(handle);
+
+    return ZJS_UNDEFINED;
+}
+
+static ZJS_DECL_FUNC(zjs_aio_open)
+{
+    // args: initialization object
+    ZJS_VALIDATE_ARGS(Z_OBJECT);
+
+    jerry_value_t data = argv[0];
+
+    u32_t pin;
+    if (!zjs_obj_get_uint32(data, "pin", &pin))
+        return zjs_error("missing required field (pin)");
+
+    if (pin < AIO_MIN || pin > AIO_MAX) {
+        DBG_PRINT("PIN: #%u\n", pin);
+        return zjs_error("pin out of range");
+    }
+
+    // create the AIOPin object
+    jerry_value_t pinobj = zjs_create_object();
+    jerry_set_prototype(pinobj, zjs_aio_prototype);
+    zjs_obj_add_number(pinobj, "pin", pin);
+
+    aio_handle_t *handle = aio_alloc_handle();
+    if (!handle) {
+        return zjs_error("could not allocate handle");
+    }
+
+    // lookup what device the pins are connected to
+    // A0 and A1 are connected to ADC_0, A2 and A3 are connected to ADC_1
+    struct device *dev = adc_device[map_device(pin)];
+    if (!dev) {
+        const char *dev_name = map_device_name(pin);
+        if (dev_name) {
+            dev = device_get_binding(dev_name);
+            if (!dev) {
+                zjs_free(handle);
+                return zjs_error("failed to initialize AIO device");
+            }
+            adc_enable(dev);
+            adc_device[map_device(pin)] = dev;
+        } else {
+            zjs_free(handle);
+            return zjs_error("failed to find ADC device");
+        }
+    }
+
+    pin_enabled[pin - AIO_MIN]++;
+    DBG_PRINT("AIO pin %d refcount %d\n", pin, pin_enabled[pin-12]);
+
+    handle->pin_obj = jerry_acquire_value(pinobj);
+    handle->pin = pin;
+
+    // make it an emitter object
+    zjs_make_emitter(pinobj, zjs_aio_prototype, handle, aio_free_cb);
+
+    // add to the list of opened handles
+    ZJS_LIST_APPEND(aio_handle_t, opened_handles, handle);
+
+    return pinobj;
+}
+
+static s32_t aio_poll_routine(void *h)
+{
+    u32_t uptime = k_uptime_get_32();
+
+    for (int i = 0; i < AIO_LEN; i++) {
+        if (pin_enabled[i]) {
+            // FIXME: We dont know when a object subscribes to onchange
+            // events, need a way to intercept the .on()
+            // currently, if the pin is opened, then it do a read
+            // and signal the event, this could slow down the system
+            // when there are multiple AIO pins opened.
+            if (uptime % (u32_t)(CONFIG_SYS_CLOCK_TICKS_PER_SEC /
+                AIO_POLL_FREQUENCY * 10) == 0) {
```

Comment:
Does this get run every millisecond? What if we don't happen to run right at the right uptime? Seems like instead you should save the last update time in a static variable or something and see if we're beyond the next poll time now?